Add Causal Discovery Problem

pacowong opened the pull request:
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue
This pull request adds a new experiment to use Nevergrad to solve the representation learning problem using Nevergrad. Specifically, it recovers a causal graph structure from data points.

## How Has This Been Tested (if it applies)
1. Test the initialization of a problem via data loader
2. Test the initialization of a problem via graph generator
3. Check the solution of the graph recovered from the graph generator

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [ ] All tests passed, and additional code has been covered with new tests.

<!--- In any case, don't hesitate to join and ask questions if you need on Nevergrad users Facebook group https://www.facebook.com/groups/nevergradusers/ -->

pacowong commented:
For your information, the following behavior is the same as that in my machine. However, the number of workers in NGO should be 1.

```python
=================================== FAILURES ===================================
____________________ test_causal_discovery_using_generator _____________________

    def test_causal_discovery_using_generator() -> None:
        nnodes = 13
        npoints = 55
        func = core.CausalDiscovery(generator='acylicgraph', npoints=npoints, nodes=nnodes)
        assert func._nvars == nnodes
        assert func._data.shape == (npoints, nnodes)
        assert func.graph_score(func._ground_truth_graph) == 1
        optimizer = ng.optimizers.NGO(parametrization=func.parametrization, budget=500)
        recommendation = optimizer.minimize(func.function)
>       assert func.objective(**recommendation.kwargs) == func._min_score_so_far #Why sometimes False?
E       assert 84.94378698224853 == 34.54212083463096
E         +84.94378698224853
E         -34.54212083463096

nevergrad/functions/causaldiscovery/test_core.py:23: AssertionError
=============================== warnings summary ===============================
```

jrapin commented on requirements/bench.txt:
```diff
@@ -14,3 +14,4 @@ torchvision
 pyomo>=5.7
 IOHexperimenter>=0.2.8
 mixsimulator>=0.2.9.5
+cdt
```
can you please add the current version as the minimal version to use? this is sometimes useful to identify easily which was a working version when code breaks

jrapin commented on nevergrad/functions/causaldiscovery/test_core.py:
```diff
@@ -0,0 +1,24 @@
+    assert func._data.shape == (npoints, nnodes)
+    assert func.graph_score(func._ground_truth_graph) == 1
+    optimizer = ng.optimizers.NGO(parametrization=func.parametrization, budget=500)
+    recommendation = optimizer.minimize(func.function)
```
this should be minimizing func, not func.function

jrapin replied:
if it was not for the check of min score so far below(which is only for debugging?), I would have said that there is no need to check that we can perform optimization here.
When implementing functions, I would rather verify that the output of an evaluation is correct, aka test an actual output of func directly.
This is particularly useful to avoid regressions, because this way we can make sure the function does not change, since we have a tests that verifies against an actual value.

Can you please add such a test?

pacowong replied:
Added `assert math.isclose(17.425619834710744, result, abs_tol=1e-10)` in `test_causal_discovery_using_data` as the graph will not change in this case.

jrapin commented on nevergrad/functions/causaldiscovery/test_core.py:
```diff
@@ -0,0 +1,24 @@
+    assert func.graph_score(func._ground_truth_graph) == 1
+    optimizer = ng.optimizers.NGO(parametrization=func.parametrization, budget=500)
+    recommendation = optimizer.minimize(func.function)
+    assert func.objective(**recommendation.kwargs) == func._min_score_so_far #Why sometimes False?
```
good question, i this is flaky we need to know why.
The answer may come from the parametrization, if it includes Choice variables, then it's noisy, so NGO may choose a noisy optimizer, which may provide an average for best recommendation. You should probably use OnePlusOne instead to avoid any issue.

teytaud replied:
NGOpt has changed quite a bit recently about these Choice variables, so maybe today the situation is quite different.

jrapin reviewed (changes requested):
You'll need to add networkx and cdt to the list of ignored packages for type checking:
https://github.com/facebookresearch/nevergrad/blob/225ff3d72b3bc32b235634bbfb8a6a10b97495de/mypy.ini#L4

teytaud commented:
@pacowong You can remove register_initialization, as in the current code this is automatically dealt with in a transparent manner.

pacowong commented:
I have removed register_initialization and updated to the latest version of Nevergrad.
But I fail the following test case:
```python
name = 'yabbob', maker = <function yabbob at 0x7f3987b16050>

    @testing.parametrized(**{name: (name, maker) for name, maker in experiments.registry.items()})
    def test_experiments_registry(name: str, maker: tp.Callable[[], tp.Iterator[experiments.Experiment]]) -> None:
        with datasets.mocked_data():  # mock mlda data that should be downloaded
>           check_maker(maker)  # this is to extract the function for reuse if other external packages need it

```

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,94 @@
+                 parents_max: int = 5,
+                 expected_degree: int = 3,
+                 ) -> None:
+        params = {x: y for x, y in locals().items() if x not in ["self", "__class__"]}  # for copying
```
```suggestion
```

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,94 @@
+        param_links = ng.p.Choice([-1, 0, 1], repetitions=self._nvars*(self._nvars-1)//2)
+        instru = ng.p.Instrumentation(network_links=param_links).set_name("")
+        super().__init__(self.objective, instru)
+        self._descriptors.update(generator=generator)
```
```suggestion
```
automatically addded

jrapin reviewed (commented):
Can you paste the error ? The one in CircleCI is different and related to the optimization I assume, while what you are pointing at seems completely disconnected

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,89 @@
+import typing as tp
```
please add the standard header for nevergrad

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,89 @@
+import typing as tp
+import networkx as nx
+import cdt
+from cdt.metrics import (precision_recall, SHD)
```
```suggestion
from cdt.metrics import precision_recall, SHD
```

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,89 @@
+from ..base import ExperimentFunction
+
+
+# pylint: disable=too-many-instance-attributes,too-many-arguments,too-many-statements,too-many-locals
```
```suggestion
# pylint: disable=too-many-instance-attributes,too-many-arguments,too-many-locals
```
I do not enforce pylint strictly but indeed I use it on my laptop. Those tags deactivate some of the issues, but I doubt you would get too-many-statements here

jrapin commented on nevergrad/functions/causaldiscovery/core.py:
```diff
@@ -0,0 +1,89 @@
+    causal_mechanism: str
+        'linear', 'polynomial', 'sigmoid_add', 'sigmoid_mix', 'gp_add', 'gp_mix', 'nn'
+    noise: str or function
+        Type of noise function to use in the generative process
```
FYI allowing functions will not allow recording it as a descriptor, which are used for the benchmarks. It is not strictly a problem, but it may get different experiments mixed up without it

jrapin commented on nevergrad/functions/causaldiscovery/test_core.py:
```diff
@@ -0,0 +1,35 @@
+    result = func(network_links=param_links.value)
+    assert np.isclose(17.425619834710744, result, atol=1e-10)
+
+    #Optimization should return the same result since the true graph is not random and small
```
can you make this a second different test?

jrapin reviewed (commented):
please see the comments and run black on the code (this is a new requirement to make sure everything is formatted the same way) and we can merge ;)

Running black: `pip install black` if you do not have it, then from the repository (the folder containing setup.py), run `black nevergrad` and you should be good

pacowong commented:
Thank you for your suggestion. The code is modified accordingly.

jrapin commented:
can you merge master in your branch? I see a lot of unrelated changes from other PRs, so it's difficult to review

pacowong commented:
@jrapin done!

jrapin reviewed (approved):
looks good to me
@teytaud good for you too? there may be some issues to solve with tensorflow version after merging, I thought that was on master

teytaud commented:
Just fixed conflicts ==> good to merge if tests ok.

teytaud commented:
@pacowong We can add this in experriments.py, right ?  so that it's now included in the dash board and all tests.

teytaud commented:
> @pacowong We can add this in experriments.py, right ? so that it's now included in the dash board and all tests.

Ok I think I'm not allowed to do a PR in your code so can you add
<< from nevergrad.functions.causaldiscovery import CausalDiscovery >> in nevergrad/benchmark/experiments.py

and, in the same file, add something like:
@registry.register
def causal_similarity(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
    """Optimizing images: artificial criterion for now."""
    seedg = create_seed_generator(seed)
    optims = ["CMA", "NGOpt8", "DE", "PSO", "RecES", "RecMixES", "RecMutDE", "ParametrizationDE"]
    if default_optims is not None:
        optims = default_optims
    func = CausalDisccovery()
    for budget in [100 * 5 ** k for k in range(3)]:
        for num_workers in [1]:
            for algo in optims:
                xp = Experiment(func, algo, budget, num_workers=num_workers, seed=next(seedg))
                if not xp.is_incoherent:
                    yield xp

pacowong commented:
> 
> 
> > @pacowong We can add this in experriments.py, right ? so that it's now included in the dash board and all tests.
> 
> Ok I think I'm not allowed to do a PR in your code so can you add
> << from nevergrad.functions.causaldiscovery import CausalDiscovery >> in nevergrad/benchmark/experiments.py
> 
> and, in the same file, add something like:
> 
> ```python
> @registry.register
> def causal_similarity(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
>     """Optimizing images: artificial criterion for now."""
>     seedg = create_seed_generator(seed)
>     optims = ["CMA", "NGOpt8", "DE", "PSO", "RecES", "RecMixES", "RecMutDE", "ParametrizationDE"]
>     func = CausalDisccovery()
>     for budget in [100 * 5 ** k for k in range(3)]:
>           for algo in optims:
>               xp = Experiment(func, algo, budget, num_workers=1, seed=next(seedg))
>               if not xp.is_incoherent:
>                   yield xp
> ```

@teytaud Done

jrapin commented on nevergrad/benchmark/experiments.py:
```diff
@@ -1486,7 +1487,21 @@ def pbo_suite(seed: tp.Optional[int] = None) -> tp.Iterator[Experiment]:
+    """Finding the best causal graph"""
+    seedg = create_seed_generator(seed)
+    optims = ["CMA", "NGOpt8", "DE", "PSO", "RecES", "RecMixES", "RecMutDE", "ParametrizationDE"]
+    if default_optims is not None:
+        optims = default_optims
```
```suggestion
```

jrapin commented:
Merged, thanks @pacowong
